smoke tests: add remix support

```
[mailing] caught error emails/TextEmail.tsx:20:8 - error TS7016: Could not find a declaration file for module 'mjml-react'. '/Users/alexfarrill/projects/mailing-test/remix-app/node_modules/mjml-react/dist/src/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/mjml-react` if it exists or add a new declaration (.d.ts) file containing `declare module 'mjml-react';`

20 } from "mjml-react";
```

--- a/.gitignore
+++ b/.gitignore
@@ -111,6 +111,7 @@ emails
 !packages/cli/src/generator_templates/js/emails/
 
 scripts/e2e_test/runs/*
+scripts/e2e_test/cache/*
 
 # Mac
 *.DS_Store

--- a/packages/cli/README.md
+++ b/packages/cli/README.md
@@ -250,3 +250,7 @@ The script supports some basic options for running today
 
 - `--only=redwood_ts` to run the tests only on the specified framework
 - `--skip-build` to skip the yarn build part of the script, useful when debugging something unrelated to the build
+
+**Cache the framework installs for faster runs**
+
+- In the directory `scripts/e2e_test`, run `bundle exec ruby e2e_test.rb --save-cache --skip-build` to save each framework install (before mailing is added) to the `cache` directory. Subsequent test runs will use the cache instead of running `yarn create` and `yarn install`, which will speed things up 🏎

--- a/scripts/e2e_test/app_configs/app.rb
+++ b/scripts/e2e_test/app_configs/app.rb
@@ -1,18 +1,18 @@
 class App
   CACHE_DIR = File.expand_path(__dir__ + '/../cache')
 
-  def initialize(name, root_dir)
+  def initialize(name, root_dir, opts)
     @name = name
     @root_dir = root_dir
+    @save_cache = opts[:save_cache]
   end
 
   def setup!
-    announce! "Creating next #{@name} app in #{@root_dir}", "⚙️"
+    announce! "Creating new #{@name} app in #{@root_dir}", "⚙️"
     FileUtils.mkdir_p(@root_dir)
 
     use_cache do
       yarn_create!
-      # prompt to save to cache?
     end
 
     verify!
@@ -29,6 +29,12 @@ private
       FileUtils.cp_r(framework_cache_dir + '/.', @root_dir)
     else
       block.call
+      # prompt to save to cache?
+
+      if @save_cache
+        verify!
+        FileUtils.cp_r(@root_dir, CACHE_DIR)
+      end
     end
   end
 

--- a/scripts/e2e_test/app_configs/app.rb
+++ b/scripts/e2e_test/app_configs/app.rb
@@ -33,7 +33,6 @@ private
       FileUtils.cp_r(framework_cache_dir + '/.', @root_dir)
     else
       block.call
-      # prompt to save to cache?
 
       if @save_cache
         verify!

--- a/scripts/e2e_test/app_configs/app.rb
+++ b/scripts/e2e_test/app_configs/app.rb
@@ -13,8 +13,9 @@ class App
     use_cache do
       yarn_create!
       # prompt to save to cache?
-      verify!
     end
+
+    verify!
 
     yalc_add_mailing!
     run_mailing!
@@ -26,8 +27,8 @@ private
     if Dir.exist?(framework_cache_dir)
       puts "Using cache..."
       FileUtils.cp_r(framework_cache_dir + '/.', @root_dir)
-      
-      verify!
+    else
+      block.call
     end
   end
 

--- a/scripts/e2e_test/app_configs/app.rb
+++ b/scripts/e2e_test/app_configs/app.rb
@@ -1,3 +1,5 @@
+require 'net/http'
+
 class App
   CACHE_DIR = File.expand_path(__dir__ + '/../cache')
 
@@ -73,7 +75,8 @@ private
     # If we can speed up the uncached previews.json load then this wait can likely be removed
     # see: https://github.com/sofn-xyz/mailing/issues/102
 
-    curl_exit_code = system_quiet("curl http://localhost:3883/previews.json")
-    fail "curl http://localhost:3883/previews.json did not succeed" unless curl_exit_code
+    uri = URI('http://localhost:3883/previews.json')
+    res = Net::HTTP.get_response(uri)
+    fail "HTTP Get #{uri} did not succeed" unless '200' == res.code
   end
 end

--- a/scripts/e2e_test/app_configs/app.rb
+++ b/scripts/e2e_test/app_configs/app.rb
@@ -3,6 +3,8 @@ require 'net/http'
 class App
   CACHE_DIR = File.expand_path(__dir__ + '/../cache')
 
+  attr_reader :io
+
   def initialize(name, root_dir, opts)
     @name = name
     @root_dir = root_dir

--- a/scripts/e2e_test/app_configs/app.rb
+++ b/scripts/e2e_test/app_configs/app.rb
@@ -1,4 +1,6 @@
 class App
+  CACHE_DIR = File.expand_path(__dir__ + '/../cache')
+
   def initialize(name, root_dir)
     @name = name
     @root_dir = root_dir
@@ -8,14 +10,27 @@ class App
     announce! "Creating next #{@name} app in #{@root_dir}", "⚙️"
     FileUtils.mkdir_p(@root_dir)
 
-    yarn_create!
-    verify!
+    use_cache do
+      yarn_create!
+      # prompt to save to cache?
+      verify!
+    end
 
     yalc_add_mailing!
     run_mailing!
   end
 
 private
+  def use_cache(&block)
+    framework_cache_dir = File.join(CACHE_DIR, @name)
+    if Dir.exist?(framework_cache_dir)
+      puts "Using cache..."
+      FileUtils.cp_r(framework_cache_dir + '/.', @root_dir)
+      
+      verify!
+    end
+  end
+
   def verify!
     fail "missing package.json in #{@root_dir}" unless File.exist?(File.join(@root_dir, "package.json"))
   end

--- a/scripts/e2e_test/app_configs/next_js_app.rb
+++ b/scripts/e2e_test/app_configs/next_js_app.rb
@@ -3,8 +3,8 @@ require_relative '../helpers/test_runner_utils'
 class NextJsApp < App
   include TestRunnerUtils
 
-  def initialize(root_dir)
-    super('next_js', root_dir)
+  def initialize(root_dir, *args)
+    super('next_js', root_dir, *args)
   end
 
 private

--- a/scripts/e2e_test/app_configs/next_ts_app.rb
+++ b/scripts/e2e_test/app_configs/next_ts_app.rb
@@ -3,8 +3,8 @@ require_relative '../helpers/test_runner_utils'
 class NextTsApp < App
   include TestRunnerUtils
 
-  def initialize(root_dir)
-    super('next_ts', root_dir)
+  def initialize(root_dir, *args)
+    super('next_ts', root_dir, *args)
   end
 
 private

--- a/scripts/e2e_test/app_configs/redwood_js_app.rb
+++ b/scripts/e2e_test/app_configs/redwood_js_app.rb
@@ -3,8 +3,8 @@ require_relative '../helpers/test_runner_utils'
 class RedwoodJsApp < App
   include TestRunnerUtils
 
-  def initialize(root_dir)
-    super('redwood_js', root_dir)
+  def initialize(root_dir, *args)
+    super('redwood_js', root_dir, *args)
   end
 
 private

--- a/scripts/e2e_test/app_configs/redwood_ts_app.rb
+++ b/scripts/e2e_test/app_configs/redwood_ts_app.rb
@@ -3,8 +3,8 @@ require_relative '../helpers/test_runner_utils'
 class RedwoodTsApp < App
   include TestRunnerUtils
 
-  def initialize(root_dir)
-    super('redwood_ts', root_dir)
+  def initialize(root_dir, *args)
+    super('redwood_ts', root_dir, *args)
   end
 
 private

--- a/scripts/e2e_test/app_configs/remix_js_app.rb
+++ b/scripts/e2e_test/app_configs/remix_js_app.rb
@@ -3,8 +3,8 @@ require_relative '../helpers/test_runner_utils'
 class RemixJsApp < App
   include TestRunnerUtils
 
-  def initialize(root_dir)
-    super('remix_js', root_dir)
+  def initialize(root_dir, *args)
+    super('remix_js', root_dir, *args)
   end
 
 private

--- a/scripts/e2e_test/app_configs/remix_js_app.rb
+++ b/scripts/e2e_test/app_configs/remix_js_app.rb
@@ -1,0 +1,16 @@
+require_relative '../helpers/test_runner_utils'
+
+class RemixJsApp < App
+  include TestRunnerUtils
+
+  def initialize(root_dir)
+    super('remix_js', root_dir)
+  end
+
+private
+  def yarn_create!
+    Dir.chdir(@root_dir) do
+      system_quiet("yarn create remix . --template=remix --no-typescript --install")
+    end
+  end
+end

--- a/scripts/e2e_test/app_configs/remix_ts_app.rb
+++ b/scripts/e2e_test/app_configs/remix_ts_app.rb
@@ -1,0 +1,16 @@
+require_relative '../helpers/test_runner_utils'
+
+class RemixTsApp < App
+  include TestRunnerUtils
+
+  def initialize(root_dir)
+    super('remix_ts', root_dir)
+  end
+
+private
+  def yarn_create!
+    Dir.chdir(@root_dir) do
+      system_quiet("yarn create remix . --template=remix --typescript --install")
+    end
+  end
+end

--- a/scripts/e2e_test/app_configs/remix_ts_app.rb
+++ b/scripts/e2e_test/app_configs/remix_ts_app.rb
@@ -3,8 +3,8 @@ require_relative '../helpers/test_runner_utils'
 class RemixTsApp < App
   include TestRunnerUtils
 
-  def initialize(root_dir)
-    super('remix_ts', root_dir)
+  def initialize(root_dir, *args)
+    super('remix_ts', root_dir, *args)
   end
 
 private

--- a/scripts/e2e_test/e2e_test.rb
+++ b/scripts/e2e_test/e2e_test.rb
@@ -59,7 +59,7 @@ class TestRunner
       begin
         tmp_dir_name = File.join(runs_dir_name, config_name.to_s)
 
-        klass.new(tmp_dir_name).setup!
+        klass.new(tmp_dir_name, save_cache: opt?('save-cache')).setup!
 
         run_cypress_tests
       ensure

--- a/scripts/e2e_test/e2e_test.rb
+++ b/scripts/e2e_test/e2e_test.rb
@@ -9,6 +9,8 @@ require_relative 'app_configs/next_ts_app'
 require_relative 'app_configs/next_js_app'
 require_relative 'app_configs/redwood_ts_app'
 require_relative 'app_configs/redwood_js_app'
+require_relative 'app_configs/remix_ts_app'
+require_relative 'app_configs/remix_js_app'
 
 class TestRunner
   include TestRunnerUtils
@@ -23,7 +25,9 @@ class TestRunner
     next_ts: NextTsApp,
     next_js: NextJsApp,
     redwood_ts: RedwoodTsApp,
-    redwood_js: RedwoodJsApp
+    redwood_js: RedwoodJsApp,
+    remix_ts: RemixTsApp,
+    remix_js: RemixJsApp
   }
 
   def initialize
@@ -135,7 +139,7 @@ private
     dirs_to_cleanup = Array(Dir.glob("#{RUNS_DIR}/*")).sort{|a,b| b <=> a}.reject{|f| f =~ /latest/}[NUM_RUNS_TO_KEEP..-1]
     dirs_to_cleanup.each do |dir|
       puts "Cleaning up #{dir}"
-      FileUtils.rm_rf(dir)
+      spawn("rm -rf #{dir}")
     end
   end
 end

--- a/scripts/e2e_test/e2e_test.rb
+++ b/scripts/e2e_test/e2e_test.rb
@@ -59,7 +59,10 @@ class TestRunner
       begin
         tmp_dir_name = File.join(runs_dir_name, config_name.to_s)
 
-        klass.new(tmp_dir_name, save_cache: opt?('save-cache')).setup!
+        app = klass.new(tmp_dir_name, save_cache: opt?('save-cache'))
+        app.setup!
+
+        @io = app.io
 
         run_cypress_tests
       ensure
@@ -125,7 +128,7 @@ private
   #
 
   def cleanup_io_and_subprocess
-    return unless @io
+    fail "No subprocess found to cleanup" unless @io
 
     # kill the subprocess
     Process.kill "INT", @io.pid

--- a/scripts/e2e_test/helpers/test_runner_utils.rb
+++ b/scripts/e2e_test/helpers/test_runner_utils.rb
@@ -10,4 +10,5 @@ module TestRunnerUtils
     puts "\n" * 10
     puts "#{emoji}  " * 10 + "\n" + text + "\n" + "#{emoji}  " * 10
   end
+  
 end

